devito/types/tensor.py

```diff
@@ -79,211 +125,89 @@ def __subfunc_setup__(cls, *args, **kwargs):
+    def __getattr__(self, name):
         """
-        Right multiplication of a TensorFunction T with u (u*T)
-        where u can be a TensorFunction, a VectorFunction or a Function/scalar.
-        Computes it via __mul__ with its transpose (T^T*u^T)^T.
+        Try calling a dynamically created FD shortcut.
+
+        Notes
+        -----
+        This method acts as a fallback for __getattribute__
         """
-        other = sympy.sympify(other)
         try:
-            if other.is_Function:
-                return self._eval_scalar_mul(other)
-            elif other.is_VectorValued or other.is_TensorValued:
-                return (self.T.__mul__(other.T)).T
-            else:
-                return super(TensorFunction, self).__rmul__(other)
-        except AttributeError:
-            return super(TensorFunction, self).__mul__(other)
+            return self.applyfunc(lambda x: x if x == 0 else getattr(x, name))
+        except:
+            raise AttributeError("%r object has no attribute %r" % (self.__class__, name))
 
     def _eval_at(self, func):
         """
         Evaluate tensor at func location
         """
-        def entry(i, j):
+        def entries(i, j, func):
             return getattr(self[i, j], '_eval_at', lambda x: self[i, j])(func[i, j])
-        comps = [[entry(i, j) for i in range(self.cols)] for j in range(self.rows)]
-        to = getattr(self, 'time_order', 0)
-        func = tens_func(self, self)
-        return func(name='%s%s' % (self.name, func.name),
-                    grid=self.grid, space_order=self.space_order,
-                    components=comps, time_order=to, symmetric=self.is_symmetric,
-                    diagonal=self.is_diagonal)
-
-    @classmethod
-    def __dtype_setup__(cls, **kwargs):
-        return Function.__dtype_setup__(**kwargs)
+        entry = lambda i, j: entries(i, j, func)
+        return self._new(self.rows, self.cols, entry)
 
     @classmethod
     def __indices_setup__(cls, **kwargs):
         return Function.__indices_setup__(grid=kwargs.get('grid'),
                                           dimensions=kwargs.get('dimensions'))
 
     @property
-    def is_diagonal(self):
-        return self._is_diagonal
-
-    @property
-    def is_symmetric(self):
-        return self._is_symmetric
-
-    @property
-    def indices(self):
-        return self._indices
-
-    @property
-    def staggered(self):
-        return self._staggered
+    def is_TimeDependent(self):
+        return self._is_TimeDependent
 
     @property
     def space_dimensions(self):
-        return self.indices
+        """Spatial dimensions."""
+        return self._space_dimensions
 
     @property
-    def grid(self):
-        return self._grid
+    def space_order(self):
+        """The space order for all components."""
+        return ({a.space_order for a in self} - {None}).pop()
 
     @property
-    def name(self):
-        return self._name
+    def time_order(self):
```

Comment:
this should only be in a `*Time*` class

Reply:
Well that one is nasty. Technically could do something like `v + u` where `v` is a VectorFunction and `u` a TimeFunction which would end up having a time_order but would need costly arg checks (this one is easy but may not be) to transfer to VectorTimeFunction so it's safer IMO to have it here in case a component end up being time-dependent.

Reply:
perhaps leave a TODO then?

also, the implementation looks broken? you're accessing `.space_order` instead of `.time_order`, and you're assuming the time orders are homogeneous. Should you not explicitly check that out, and raise an Exception if that is not the case?

about promotion to VectorTimeFunction... I guess you would do that in `_new` right? I don't understand why you say it'd be costly, do you not just have to check whether at least one arg is time-depedent?

Reply:
> don't understand why you say it'd be costly, do you not just have to check whether at least one arg is time-dependent?

Then you end up with odd circular imports and dependency.

>you're accessing .space_order instead of .time_order.

That's a bug, fixed.